Can't use database schema with a dash in its name (liquibase, postgres)

### Before reporting an issue

- [X] I have searched existing issues
- [X] I have reproduced the issue with the latest release

### Area

storage

### Describe the bug

I created a clean lab environment with keycloak v21.1.1 and postgres 12, both via docker container. The database was prepared with dbname `sso-stack-lab`, schema `keycloak-lab` and a user `keycloak-lab`. Configfile `keycloak.conf` is mounted from local host into the container, setting `db=postgres`, `db-url-host=localhost`, `db-username=keycloak-lab`, `db-password=mySecretPWD`, `db-schema=keycloak-lab`

When starting keycloak (from a custom image) I get the error `liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"`

After stumbling across #17276 I changed the database name to no avail. Next I tried changing the schema name to `keycloaklab` (removing the dash) and now it works. I wasnt able to further pinpoint the problem, but after removing the dash from schema name, but with username name and database name still having a dash, it works.

### Version

21.1.1

### Expected behavior

Using a database schema name with a dash, keycloak starts and works as expected.

### Actual behavior

When starting keycloak (from a custom image with added certificates and build run done) with a database schema name including a dash I get the error `liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"` and keycloak will fail to start.

### How to Reproduce?

Run a clean install of keycloak, using postgres with a database schema name including a dash.

### Anything else?

_No response_

I was able to reproduce the issue. There is full stacktrace.
```
java.lang.RuntimeException: liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.lambda$getLiquibaseUnrunChangeSets$0(QuarkusJpaUpdaterProvider.java:299)
	at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.getLiquibaseUnrunChangeSets(QuarkusJpaUpdaterProvider.java:295)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.validateChangeSet(QuarkusJpaUpdaterProvider.java:265)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.validate(QuarkusJpaUpdaterProvider.java:237)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.createOrUpdateSchema(LegacyJpaConnectionProviderFactory.java:255)
	at org.keycloak.quarkus.runtime.storage.legacy.database.LegacyJpaConnectionProviderFactory.postInit(LegacyJpaConnectionProviderFactory.java:129)
	at org.keycloak.quarkus.runtime.integration.QuarkusKeycloakSessionFactory.init(QuarkusKeycloakSessionFactory.java:105)
	at org.keycloak.quarkus.runtime.integration.jaxrs.QuarkusKeycloakApplication.createSessionFactory(QuarkusKeycloakApplication.java:45)
	at org.keycloak.services.resources.KeycloakApplication.startup(KeycloakApplication.java:125)
	at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver.onStartupEvent(QuarkusLifecycleObserver.java:37)
	at org.keycloak.quarkus.runtime.integration.QuarkusLifecycleObserver_Observer_onStartupEvent_c9888fa6c2aa9208d4625ee9d83de6fd77e22c83.notify(Unknown Source)
	at io.quarkus.arc.impl.EventImpl$Notifier.notifyObservers(EventImpl.java:346)
	at io.quarkus.arc.impl.EventImpl$Notifier.notify(EventImpl.java:328)
	at io.quarkus.arc.impl.EventImpl.fire(EventImpl.java:82)
	at io.quarkus.arc.runtime.ArcRecorder.fireLifecycleEvent(ArcRecorder.java:155)
	at io.quarkus.arc.runtime.ArcRecorder.handleLifecycleEvents(ArcRecorder.java:106)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy_0(Unknown Source)
	at io.quarkus.deployment.steps.LifecycleEventsBuildStep$startupEvent1144526294.deploy(Unknown Source)
	at io.quarkus.runner.ApplicationImpl.doStart(Unknown Source)
	at io.quarkus.runtime.Application.start(Application.java:101)
	at io.quarkus.runtime.ApplicationLifecycleManager.run(ApplicationLifecycleManager.java:111)
	at io.quarkus.runtime.Quarkus.run(Quarkus.java:71)
	at org.keycloak.quarkus.runtime.KeycloakMain.start(KeycloakMain.java:98)
	at org.keycloak.quarkus.runtime.cli.command.AbstractStartCommand.run(AbstractStartCommand.java:37)
	at picocli.CommandLine.executeUserObject(CommandLine.java:2026)
	at picocli.CommandLine.access$1500(CommandLine.java:148)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2461)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2453)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2415)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2273)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2417)
	at picocli.CommandLine.execute(CommandLine.java:2170)
	at org.keycloak.quarkus.runtime.cli.Picocli.parseAndRun(Picocli.java:100)
	at org.keycloak.quarkus.runtime.KeycloakMain.main(KeycloakMain.java:88)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.doRun(QuarkusEntryPoint.java:61)
	at io.quarkus.bootstrap.runner.QuarkusEntryPoint.main(QuarkusEntryPoint.java:32)
Caused by: liquibase.exception.LiquibaseException: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at liquibase.Liquibase.runInScope(Liquibase.java:1623)
	at liquibase.Liquibase.listUnrunChangeSets(Liquibase.java:1279)
	at org.keycloak.quarkus.runtime.storage.legacy.liquibase.QuarkusJpaUpdaterProvider.lambda$getLiquibaseUnrunChangeSets$0(QuarkusJpaUpdaterProvider.java:297)
	... 40 more
Caused by: liquibase.exception.UnexpectedLiquibaseException: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at liquibase.changelog.StandardChangeLogHistoryService.hasDatabaseChangeLogTable(StandardChangeLogHistoryService.java:88)
	at liquibase.changelog.StandardChangeLogHistoryService.getRanChangeSets(StandardChangeLogHistoryService.java:317)
	at org.keycloak.connections.jpa.updater.liquibase.conn.CustomChangeLogHistoryService.getRanChangeSets(CustomChangeLogHistoryService.java:54)
	at liquibase.database.AbstractJdbcDatabase.getRanChangeSetList(AbstractJdbcDatabase.java:1156)
	at liquibase.changelog.DatabaseChangeLog.validate(DatabaseChangeLog.java:342)
	at liquibase.Liquibase.lambda$listUnrunChangeSets$16(Liquibase.java:1287)
	at liquibase.Scope.lambda$child$0(Scope.java:194)
	at liquibase.Scope.child(Scope.java:203)
	at liquibase.Scope.child(Scope.java:193)
	at liquibase.Scope.child(Scope.java:172)
	at liquibase.Liquibase.runInScope(Liquibase.java:1618)
	... 42 more
Caused by: liquibase.exception.DatabaseException: ERROR: syntax error at or near "-"
  Position: 22 [Failed SQL: (0) SET SEARCH_PATH TO kc-db, "$user","public"]
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:445)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:77)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:160)
	at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:125)
	at liquibase.database.core.DatabaseUtils.initializeDatabase(DatabaseUtils.java:61)
	at liquibase.database.core.PostgresDatabase.rollback(PostgresDatabase.java:403)
	at liquibase.snapshot.SnapshotGeneratorFactory.has(SnapshotGeneratorFactory.java:118)
	at liquibase.snapshot.SnapshotGeneratorFactory.hasDatabaseChangeLogTable(SnapshotGeneratorFactory.java:269)
	at liquibase.changelog.StandardChangeLogHistoryService.hasDatabaseChangeLogTable(StandardChangeLogHistoryService.java:86)
	... 52 more
Caused by: org.postgresql.util.PSQLException: ERROR: syntax error at or near "-"
  Position: 22
	at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2713)
	at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2401)
	at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:368)
	at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:498)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:415)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:335)
	at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:321)
	at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:297)
	at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:292)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.postgresql.ds.PGPooledConnection$StatementHandler.invoke(PGPooledConnection.java:441)
	at jdk.proxy2/jdk.proxy2.$Proxy75.execute(Unknown Source)
	at io.agroal.pool.wrapper.StatementWrapper.execute(StatementWrapper.java:235)
	at liquibase.executor.jvm.JdbcExecutor$ExecuteStatementCallback.doInStatement(JdbcExecutor.java:441)
	... 60 more
```

--- a/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java
+++ b/model/jpa/src/main/java/org/keycloak/connections/jpa/DefaultJpaConnectionProviderFactory.java
@@ -59,6 +59,7 @@ import java.util.Collection;
 import java.util.HashMap;
 import java.util.LinkedHashMap;
 import java.util.Map;
+import liquibase.GlobalConfiguration;
 
 /**
  * @author <a href="mailto:sthorger@redhat.com">Stian Thorgersen</a>
@@ -434,7 +435,12 @@ public class DefaultJpaConnectionProviderFactory implements JpaConnectionProvide
 
     @Override
     public String getSchema() {
-        return config.get("schema");
+        String schema = config.get("schema");
+        if (schema != null && schema.contains("-") && ! Boolean.parseBoolean(System.getProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey()))) {
+            System.setProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey(), "true");
+            logger.warnf("The passed schema '%s' contains a dash. Setting liquibase config option PRESERVE_SCHEMA_CASE to true. See https://github.com/keycloak/keycloak/issues/20870 for more information.", schema);
+        }
+        return schema;
     }
 
     @Override

--- a/model/jpa/src/main/java/org/keycloak/connections/jpa/util/JpaUtils.java
+++ b/model/jpa/src/main/java/org/keycloak/connections/jpa/util/JpaUtils.java
@@ -34,7 +34,6 @@ import java.io.IOException;
 import java.io.InputStream;
 import java.net.URL;
 import java.util.ArrayList;
-import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.Properties;
@@ -52,7 +51,7 @@ public class JpaUtils {
 
     public static String getTableNameForNativeQuery(String tableName, EntityManager em) {
         String schema = (String) em.getEntityManagerFactory().getProperties().get(HIBERNATE_DEFAULT_SCHEMA);
-        return (schema==null) ? tableName : schema + "." + tableName;
+        return (schema==null) ? tableName : "\"" + schema + "\"." + tableName;
     }
 
     public static EntityManagerFactory createEntityManagerFactory(KeycloakSession session, String unitName, Map<String, Object> properties, boolean jta) {

--- a/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java
+++ b/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapStorageProviderFactory.java
@@ -46,6 +46,7 @@ import jakarta.transaction.NotSupportedException;
 import jakarta.transaction.RollbackException;
 import jakarta.transaction.SystemException;
 import jakarta.transaction.Transaction;
+import liquibase.GlobalConfiguration;
 
 import org.hibernate.cfg.AvailableSettings;
 import org.hibernate.internal.SessionImpl;
@@ -401,7 +402,7 @@ public class JpaMapStorageProviderFactory implements
             }
         }
 
-        String schema = config.get("schema");
+        String schema = getSchema();
         if (schema != null) {
             properties.put(HIBERNATE_DEFAULT_SCHEMA, schema);
         }
@@ -428,6 +429,15 @@ public class JpaMapStorageProviderFactory implements
         logger.trace("EntityManagerFactory created");
 
         return emf;
+    }
+
+    private String getSchema() {
+        String schema = config.get("schema");
+        if (schema != null && schema.contains("-") && ! Boolean.parseBoolean(System.getProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey()))) {
+            System.setProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey(), "true");
+            logger.warnf("The passed schema '%s' contains a dash. Setting liquibase config option PRESERVE_SCHEMA_CASE to true. See https://github.com/keycloak/keycloak/issues/20870 for more information.", schema);
+        }
+        return schema;
     }
 
     protected EntityManagerFactory getEntityManagerFactory() {
@@ -464,7 +474,7 @@ public class JpaMapStorageProviderFactory implements
                             if (logger.isDebugEnabled()) printOperationalInfo(connection);
 
                             MapJpaUpdaterProvider updater = session.getProvider(MapJpaUpdaterProvider.class);
-                            MapJpaUpdaterProvider.Status status = updater.validate(modelType, connection, config.get("schema"));
+                            MapJpaUpdaterProvider.Status status = updater.validate(modelType, connection, getSchema());
                             databaseShortName = updater.getDatabaseShortName();
 
                             if (!status.equals(VALID)) {
@@ -544,10 +554,10 @@ public class JpaMapStorageProviderFactory implements
     private void update(Class<?> modelType, Connection connection, KeycloakSession session) {
         if (modelType == MapLockEntity.class) {
             // as the MapLockEntity is used by the MapGlobalLockProvider itself, don't create a global lock for creating that schema
-            session.getProvider(MapJpaUpdaterProvider.class).update(modelType, connection, config.get("schema"));
+            session.getProvider(MapJpaUpdaterProvider.class).update(modelType, connection, getSchema());
         } else {
             session.getProvider(GlobalLockProvider.class).withLock(modelType.getName(), lockedSession -> {
-                lockedSession.getProvider(MapJpaUpdaterProvider.class).update(modelType, connection, config.get("schema"));
+                lockedSession.getProvider(MapJpaUpdaterProvider.class).update(modelType, connection, getSchema());
                 return null;
             });
         }

--- a/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapUtils.java
+++ b/model/map-jpa/src/main/java/org/keycloak/models/map/storage/jpa/JpaMapUtils.java
@@ -26,7 +26,6 @@ import jakarta.persistence.EntityManagerFactory;
 import java.io.IOException;
 import java.io.InputStream;
 import java.net.URL;
-import java.util.Collections;
 import java.util.Map;
 import java.util.Properties;
 import java.util.regex.Pattern;
@@ -44,7 +43,7 @@ public class JpaMapUtils {
 
     public static String getSchemaForNativeQuery(EntityManager em) {
         String schema = (String) em.getEntityManagerFactory().getProperties().get(HIBERNATE_DEFAULT_SCHEMA);
-        return (schema == null) ? "" : schema + ".";
+        return (schema == null) ? "" : "\"" + schema + "\".";
     }
 
     /**

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/database/jpa/AbstractJpaConnectionProviderFactory.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/storage/database/jpa/AbstractJpaConnectionProviderFactory.java
@@ -36,8 +36,12 @@ import org.keycloak.quarkus.runtime.configuration.Configuration;
 
 import io.quarkus.arc.Arc;
 import io.quarkus.hibernate.orm.PersistenceUnit;
+import liquibase.GlobalConfiguration;
+import org.jboss.logging.Logger;
 
 public abstract class AbstractJpaConnectionProviderFactory implements JpaConnectionProviderFactory {
+
+    private final Logger logger = Logger.getLogger(getClass());
 
     protected Config.Scope config;
     protected EntityManagerFactory entityManagerFactory;
@@ -55,7 +59,12 @@ public abstract class AbstractJpaConnectionProviderFactory implements JpaConnect
 
     @Override
     public String getSchema() {
-        return Configuration.getRawValue("kc.db-schema");
+        String schema = Configuration.getRawValue("kc.db-schema");
+        if (schema != null && schema.contains("-") && ! Boolean.parseBoolean(System.getProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey()))) {
+            System.setProperty(GlobalConfiguration.PRESERVE_SCHEMA_CASE.getKey(), "true");
+            logger.warnf("The passed schema '%s' contains a dash. Setting liquibase config option PRESERVE_SCHEMA_CASE to true. See https://github.com/keycloak/keycloak/issues/20870 for more information.", schema);
+        }
+        return schema;
     }
 
     @Override

